packages/flutter/lib/src/services/spell_check.dart

```diff
@@ -0,0 +1,553 @@
+    static SpellCheckConfiguration disabled = SpellCheckConfiguration();
+
+    static SpellCheckService? getDefaultSpellCheckService(TargetPlatform platform) {
+        switch(platform) {
+            case TargetPlatform.android:
+                return MaterialSpellCheckService();
+            default:
+                return null;
+
+        }
+    }
+}
+
+/// Interface that represents the core functionality needed to support spell check on text input.
+abstract class SpellCheckService {
+    // Initiates spell check. Expected to set spellCheckSuggestions in handler if synchronous.
+    Future<List<dynamic>> fetchSpellCheckSuggestions(Locale locale, TextEditingValue value);
+}
+
+/// Interface that represents the core functionality needed to display results of spell check.
+abstract class SpellCheckSuggestionsHandler {
+    // Builds toolbar/menu that will display spell check results.
+    Widget buildSpellCheckSuggestionsToolbar(
+        List<SpellCheckerSuggestionSpan>? spellCheckResults,
+        TextSelectionDelegate delegate, 
+        List<TextSelectionPoint> endpoints, Rect globalEditableRegion, 
+        Offset selectionMidpoint, double textLineHeight);
+
+    // Build TextSpans with misspelled words indicated.
+    TextSpan buildTextSpanWithSpellCheckSuggestions(
+        List<SpellCheckerSuggestionSpan>? spellCheckResults, String? spellCheckResultsText,
+        TextEditingValue value, TextStyle? style, bool composingWithinCurrentTextRange);
+}
+
+class DefaultSpellCheckSuggestionsHandler implements SpellCheckSuggestionsHandler {
+    //TODO(camillesimon): Replace method of building TextSpan tree in three parts with method of building in one part. Attempt started below.
+    int scssSpans_consumed_index = 0;
+    int text_consumed_index = 0;
+
+    String lastUsedText = "";
+
+    final TargetPlatform platform;
+
+    DefaultSpellCheckSuggestionsHandler(this.platform);
+
+    @override
+    Widget buildSpellCheckSuggestionsToolbar(
+        List<SpellCheckerSuggestionSpan>? spellCheckResults,
+        TextSelectionDelegate delegate, 
+        List<TextSelectionPoint> endpoints, Rect globalEditableRegion, 
+        Offset selectionMidpoint, double textLineHeight) {
+            return _SpellCheckerSuggestionsToolbar(
+            platform: platform,
+            delegate: delegate,
+            endpoints: endpoints,
+            globalEditableRegion: globalEditableRegion,
+            selectionMidpoint: selectionMidpoint,
+            textLineHeight: textLineHeight,
+            spellCheckerSuggestionSpans: spellCheckResults,
+            );
+        }
+
+   // Provides a generous guesss of the spell check results for the current text if the spell check results for this text has not been received by the framework yet.
+   // Assumes: [1] order of results matches that of the text, [2] only a shift/deletion occurs at a time (this is verifiable)
```

Comment:
Does this really assume that the difference between newText and resultsText is only a shift or deletion?  It looks like it would work (at least partially) in other cases, it's just best in the case of shifts and deletions.

Reply:
Ah I think you're right! I just meant that we can only guarantee accuracy for shifts or deletions because otherwise we just throw out the spell check result since this algorithm won't find it.